File: nothing/capture.py
"""Screenshot and screen recording for Nothing phones."""

import subprocess
from datetime import datetime
from pathlib import Path

from .device import adb_pull, run
from .exceptions import AdbError
from .models import DeviceInfo

_MAX_RECORD_DURATION = 180  # seconds; hard limit imposed by screenrecord


def _timestamp() -> str:
    return datetime.now().strftime("%Y%m%d_%H%M%S")


def action_screenshot(device: DeviceInfo, base_dir: Path) -> None:
    """Capture a screenshot and pull it to base_dir/screenshots/."""
    ts = _timestamp()
    remote_path = f"/sdcard/Download/screenshot_{ts}.png"
    local_dir = base_dir / "screenshots"
    local_dir.mkdir(parents=True, exist_ok=True)
    local_path = local_dir / f"screenshot_{ts}.png"

    print(f"Taking screenshot...")
    r = run(["adb", "-s", device.serial, "shell", "screencap", "-p", remote_path])
    if r.returncode != 0:
        raise AdbError(f"screencap failed: {r.stderr.strip()}")

    try:
        adb_pull(remote_path, local_path, device.serial)
    finally:
        run(["adb", "-s", device.serial, "shell", "rm", "-f", remote_path])

    if not local_path.exists():
        raise AdbError(f"Screenshot file not found after pull: {local_path}")

    size_kb = local_path.stat().st_size / 1024
    print(f"[OK] Screenshot saved: {local_path}")
    print(f"     Size: {size_kb:.1f} KB")


def action_screenrecord(device: DeviceInfo, base_dir: Path,
                        duration: int = 30) -> None:
    """Record the screen and pull the video to base_dir/recordings/."""
    if duration > _MAX_RECORD_DURATION:
        print(f"[WARN] Requested duration {duration}s exceeds maximum "
              f"{_MAX_RECORD_DURATION}s — clamping.")
        duration = _MAX_RECORD_DURATION

    ts = _timestamp()
    remote_path = f"/sdcard/Download/screenrecord_{ts}.mp4"
    local_dir = base_dir / "recordings"
    local_dir.mkdir(parents=True, exist_ok=True)
    local_path = local_dir / f"screenrecord_{ts}.mp4"

    # Some devices (e.g. Nothing Phone 1) require an explicit --size or the
    # MediaCodec encoder fails with err=-38. Scale to 720p width to stay within
    # encoder limits while preserving the native aspect ratio.
    size_arg: list[str] = []
    r_size = run(["adb", "-s", device.serial, "shell", "wm size"])
    if r_size.returncode == 0:
        import re as _re
        m = _re.search(r"Physical size:\s*(\d+)x(\d+)", r_size.stdout)
        if m:
            w, h = int(m.group(1)), int(m.group(2))
            if w > 720:
                h = round(h * 720 / w)
                w = 720
            size_arg = ["--size", f"{w}x{h}"]

    print(f"Recording for {duration} seconds (Ctrl-C to stop early)...")

    timeout_ms = (duration + 10) * 1000
    try:
        r = subprocess.run(
            ["adb", "-s", device.serial, "shell",
             "screenrecord", "--time-limit", str(duration), *size_arg, remote_path],
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=timeout_ms / 1000,
        )
    except subprocess.TimeoutExpired:
        print("[WARN] screenrecord timed out — attempting to pull partial recording.")
        r = None
    except KeyboardInterrupt:
        print("\n[WARN] Interrupted — attempting to pull partial recording.")
        r = None

    if r is not None:
        combined = (r.stdout + r.stderr).lower()
        if r.returncode != 0 and ("not found" in combined or "screenrecord" in combined
                                  and "permission denied" not in combined
                                  and r.returncode == 127):
            raise AdbError(
                "screenrecord is not available on this device or Android version. "
                "Requires Android 4.4+ and is absent on some low-RAM or Go devices."
            )
        if r.returncode != 0 and "not found" in combined:
            raise AdbError(
                "screenrecord is not available on this device or Android version."
            )

    try:
        adb_pull(remote_path, local_path, device.serial)
    except AdbError as exc:
        raise AdbError(f"Failed to pull recording: {exc}") from exc
    finally:
        run(["adb", "-s", device.serial, "shell", "rm", "-f", remote_path])

    if not local_path.exists():
        raise AdbError(f"Recording file not found after pull: {local_path}")

    size_mb = local_path.stat().st_size / (1024 * 1024)
    print(f"[OK] Recording saved: {local_path}")
    print(f"     Size: {size_mb:.2f} MB")

File: go/internal/capture/capture.go
// Package capture provides screenshot and screen recording for Nothing phones.
package capture

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/Limplom/nothingctl/internal/adb"
	nterrors "github.com/Limplom/nothingctl/internal/errors"
)

const maxRecordDuration = 180 // seconds; hard limit imposed by screenrecord

func timestamp() string {
	return time.Now().Format("20060102_150405")
}

// ActionScreenshot captures a screenshot and pulls it to baseDir/screenshots/.
func ActionScreenshot(serial, baseDir string) error {
	ts := timestamp()
	remotePath := fmt.Sprintf("/sdcard/Download/screenshot_%s.png", ts)
	destDir := filepath.Join(baseDir, "screenshots")
	if err := os.MkdirAll(destDir, 0755); err != nil {
		return err
	}
	localPath := filepath.Join(destDir, fmt.Sprintf("screenshot_%s.png", ts))

	fmt.Println("  Taking screenshot...")
	_, stderr, code := adb.Run([]string{"adb", "-s", serial, "shell", "screencap", "-p", remotePath})
	if code != 0 {
		return nterrors.AdbError(fmt.Sprintf("screencap failed: %s", strings.TrimSpace(stderr)))
	}

	pullErr := adb.AdbPull(serial, remotePath, localPath)
	// Always clean up remote file
	adb.Run([]string{"adb", "-s", serial, "shell", "rm", "-f", remotePath})

	if pullErr != nil {
		return pullErr
	}

	info, err := os.Stat(localPath)
	if err != nil {
		return nterrors.AdbError(fmt.Sprintf("Screenshot file not found after pull: %s", localPath))
	}
	sizeKB := float64(info.Size()) / 1024
	fmt.Printf("[OK] Screenshot saved: %s\n", localPath)
	fmt.Printf("     Size: %.1f KB\n", sizeKB)
	return nil
}

var wmSizeRe = regexp.MustCompile(`Physical size:\s*(\d+)x(\d+)`)

// ActionScreenrecord records the screen and pulls the video to baseDir/recordings/.
func ActionScreenrecord(serial, baseDir string, duration int) error {
	if duration > maxRecordDuration {
		fmt.Printf("[WARN] Requested duration %ds exceeds maximum %ds — clamping.\n",
			duration, maxRecordDuration)
		duration = maxRecordDuration
	}

	ts := timestamp()
	remotePath := fmt.Sprintf("/sdcard/Download/screenrecord_%s.mp4", ts)
	destDir := filepath.Join(baseDir, "recordings")
	if err := os.MkdirAll(destDir, 0755); err != nil {
		return err
	}
	localPath := filepath.Join(destDir, fmt.Sprintf("screenrecord_%s.mp4", ts))

	// Determine --size arg to avoid encoder failures (e.g. Nothing Phone 1)
	var sizeArgs []string
	sizeOut, _, _ := adb.Run([]string{"adb", "-s", serial, "shell", "wm size"})
	if m := wmSizeRe.FindStringSubmatch(sizeOut); m != nil {
		var w, h int
		fmt.Sscanf(m[1], "%d", &w)
		fmt.Sscanf(m[2], "%d", &h)
		if w > 720 {
			h = int(float64(h) * 720 / float64(w))
			w = 720
		}
		sizeArgs = []string{"--size", fmt.Sprintf("%dx%d", w, h)}
	}

	fmt.Printf("  Recording for %d seconds (Ctrl-C to stop early)...\n", duration)

	cmdArgs := []string{"adb", "-s", serial, "shell",
		"screenrecord", "--time-limit", fmt.Sprintf("%d", duration)}
	cmdArgs = append(cmdArgs, sizeArgs...)
	cmdArgs = append(cmdArgs, remotePath)

	timeout := time.Duration(duration+10) * time.Second
	c := exec.Command(cmdArgs[0], cmdArgs[1:]...)
	var outBuf, errBuf strings.Builder
	c.Stdout = &outBuf
	c.Stderr = &errBuf

	done := make(chan error, 1)
	if err := c.Start(); err != nil {
		return nterrors.AdbError(fmt.Sprintf("could not start screenrecord: %v", err))
	}
	go func() { done <- c.Wait() }()

	var runErr error
	select {
	case runErr = <-done:
		// command finished
	case <-time.After(timeout):
		c.Process.Kill()
		fmt.Println("[WARN] screenrecord timed out — attempting to pull partial recording.")
		runErr = nil
	}

	if runErr != nil {
		combined := strings.ToLower(outBuf.String() + errBuf.String())
		if strings.Contains(combined, "not found") {
			return nterrors.AdbError(
				"screenrecord is not available on this device or Android version. " +
					"Requires Android 4.4+ and is absent on some low-RAM or Go devices.")
		}
	}

	pullErr := adb.AdbPull(serial, remotePath, localPath)
	// Always clean up remote file
	adb.Run([]string{"adb", "-s", serial, "shell", "rm", "-f", remotePath})

	if pullErr != nil {
		return nterrors.AdbError(fmt.Sprintf("Failed to pull recording: %v", pullErr))
	}

	info, err := os.Stat(localPath)
	if err != nil {
		return nterrors.AdbError(fmt.Sprintf("Recording file not found after pull: %s", localPath))
	}
	sizeMB := float64(info.Size()) / (1024 * 1024)
	fmt.Printf("[OK] Recording saved: %s\n", localPath)
	fmt.Printf("     Size: %.2f MB\n", sizeMB)
	return nil
}
